src/nunit.analyzers.tests/SameAsOnValueTypes/SameAsOnValueTypesAnalyzerTests.cs

```diff
@@ -0,0 +1,106 @@
+using Gu.Roslyn.Asserts;
+using Microsoft.CodeAnalysis.Diagnostics;
+using NUnit.Analyzers.Constants;
+using NUnit.Analyzers.SameAsOnValueTypes;
+using NUnit.Framework;
+
+namespace NUnit.Analyzers.Tests.SameAsOnValueTypes
+{
+    public class SameAsOnValueTypesAnalyzerTests
+    {
+        private static readonly DiagnosticAnalyzer analyzer = new SameAsOnValueTypesAnalyzer();
+        private static readonly ExpectedDiagnostic expectedDiagnostic =
+            ExpectedDiagnostic.Create(AnalyzerIdentifiers.SameAsOnValueTypes);
+
+        [Test]
+        public void AnalyzeWhenLiteralTypesProvided()
+        {
+            var testCode = TestUtility.WrapInTestMethod(
+                "↓Assert.That(1, Is.SameAs(1));");
+
+            AnalyzerAssert.Diagnostics(analyzer, expectedDiagnostic, testCode);
+        }
+
+        [Test]
+        public void AnalyzeWhenStructTypesProvided()
+        {
+            var testCode = TestUtility.WrapInTestMethod(@"
+                var expected = Guid.Empty;
+                var actual = expected;
+
+                ↓Assert.That(actual, Is.SameAs(expected));");
+
+            AnalyzerAssert.Diagnostics(analyzer, expectedDiagnostic, testCode);
+        }
+
+        [Test]
+        public void AnalyzeWhenStructDelegateProvided()
+        {
+            var testCode = TestUtility.WrapInTestMethod(@"
+                var expected = Guid.Empty;
+
+                ↓Assert.That(() => expected, Is.SameAs(expected));");
+
+            AnalyzerAssert.Diagnostics(analyzer, expectedDiagnostic, testCode);
+        }
+
+        [Test]
+        public void AnalyzeWhenStructTaskProvided()
+        {
+            var testCode = TestUtility.WrapInTestMethod(@"
+                var expected = Guid.Empty;
+
+                ↓Assert.That(Task.FromResult(expected), Is.SameAs(expected));");
+
+            AnalyzerAssert.Diagnostics(analyzer, expectedDiagnostic, testCode);
+        }
+
+        [Test]
+        public void AnalyzeWhenOneArgumentIsCastToObject()
+        {
+            var testCode = TestUtility.WrapInTestMethod(
+                "↓Assert.That((object)Guid.Empty, Is.Not.SameAs(Guid.Empty));");
+
+            AnalyzerAssert.Diagnostics(analyzer, expectedDiagnostic, testCode);
+        }
+
+        [Test]
+        public void AnalyzeWhenOtherArgumentIsCastToObject()
+        {
+            var testCode = TestUtility.WrapInTestMethod(
+                "↓Assert.That(Guid.Empty, Is.Not.SameAs((object)Guid.Empty));");
+
+            AnalyzerAssert.Diagnostics(analyzer, expectedDiagnostic, testCode);
+        }
+
+        [Test]
+        public void AnalyzeWhenBothArgumentsAreCastToObject()
```

Comment:
(Specifically the salient point is that both arguments are reference-typed, not that there is a cast in the expression.)